src/scale_impls/decode.rs

```diff
@@ -13,239 +13,296 @@
+        while let Some(val) = $value.decode_item(DecodeValueVisitor::<R, T, F>::new()) {
             let val = val?;
             vals.push(val);
         }
-        Ok(Value { value: ValueDef::Composite(Composite::Unnamed(vals)), context: $type_id.0 })
+        Ok(Value {
+            value: ValueDef::Composite(Composite::Unnamed(vals)),
+            context: F::map($type_id),
+        })
     }};
 }
 
 // We can't implement this on `Value<TypeId>` because we have no TypeId to assign to the value.
-impl scale_decode::DecodeAsFields for Composite<TypeId> {
-    fn decode_as_fields<'info>(
+impl scale_decode::DecodeAsFields for Composite<()> {
+    fn decode_as_fields<'resolver, R: TypeResolver>(
         input: &mut &[u8],
-        fields: &mut dyn FieldIter<'info>,
-        types: &'info PortableRegistry,
+        fields: &mut dyn FieldIter<'resolver, R::TypeId>,
+        types: &'resolver R,
     ) -> Result<Self, scale_decode::Error> {
         // Build a Composite type to pass to a one-off visitor:
-        static EMPTY_PATH: &Path<PortableForm> = &Path { segments: Vec::new() };
         let mut composite =
-            scale_decode::visitor::types::Composite::new(input, EMPTY_PATH, fields, types, false);
+            scale_decode::visitor::types::Composite::new(input, fields, types, false);
         // Decode into a Composite value from this:
-        let val = visit_composite(&mut composite);
+        let val = visit_composite::<R, (), DefaultMapper>(&mut composite);
         // Consume remaining bytes and update input cursor:
         composite.skip_decoding()?;
         *input = composite.bytes_from_undecoded();
-
-        val.map_err(From::from)
+        val.map_err(From::from).map(|v| v.map_context(|_| {}))
     }
 }
 
 /// A [`scale_decode::Visitor`] implementation for decoding into [`Value`]s.
-pub struct DecodeValueVisitor;
+pub struct DecodeValueVisitor<R: TypeResolver, T, F> {
+    resolver: PhantomData<(R, T, F)>,
+}
+impl<R: TypeResolver, T, F> Default for DecodeValueVisitor<R, T, F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<R: TypeResolver, T, F> DecodeValueVisitor<R, T, F> {
+    pub fn new() -> Self {
+        DecodeValueVisitor { resolver: PhantomData }
+    }
+}
+
+impl scale_decode::IntoVisitor for Value<()> {
+    // Note: the DefaultMapper just removes all type ids here.
+    type AnyVisitor<R: scale_decode::TypeResolver> =
+        scale_decode::visitor::VisitorWithCrateError<DecodeValueVisitor<R, (), DefaultMapper>>;
 
-impl scale_decode::IntoVisitor for Value<TypeId> {
-    type Visitor = scale_decode::visitor::VisitorWithCrateError<DecodeValueVisitor>;
-    fn into_visitor() -> Self::Visitor {
-        scale_decode::visitor::VisitorWithCrateError(DecodeValueVisitor)
+    fn into_visitor<R: scale_decode::TypeResolver>() -> Self::AnyVisitor<R> {
+        scale_decode::visitor::VisitorWithCrateError(DecodeValueVisitor::new())
```

Comment:
Here I think you could use `value.map_context` to remove the `R::TypeId` from the returned `Value<R::TypeId>`; perhaps this is what the mapper stuff is still here for, so I think it could be removed with this tweak?

Reply:
But where should this context mapping happen? Here we need to return a Visitor that always returns Value<()> so the `map_context` needs to happen in the `visit_*()` functions of that visitor, where I already use `map_context` to achieve this. Or maybe I don't understand.

Reply:
Yup, good point :)